libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -666,56 +666,134 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
 
 void ContractLevelChecker::checkAmbiguousOverrides(ContractDefinition const& _contract) const
 {
-	vector<FunctionDefinition const*> contractFuncs = _contract.definedFunctions();
-
-	auto const resolvedBases = resolveDirectBaseContracts(_contract);
-
-	FunctionMultiSet inheritedFuncs = inheritedFunctions(&_contract);;
-
-	// Check the sets of the most-inherited functions
-	for (auto it = inheritedFuncs.cbegin(); it != inheritedFuncs.cend(); it = inheritedFuncs.upper_bound(*it))
+	// Fetch inherited functions and sort them by signature.
+	// We get at least one function per signature and direct base contract, which is
+	// enough because we re-construct the inheritance graph later.
+	FunctionMultiSet nonOverriddenFunctions = inheritedFunctions(_contract);
+	// Remove all functions that match the signature of a function in the current contract.
+	nonOverriddenFunctions -= _contract.definedFunctions();
+
+	// Walk through the set of functions signature by signature.
+	for (auto it = nonOverriddenFunctions.cbegin(); it != nonOverriddenFunctions.cend();)
 	{
-		auto [begin,end] = inheritedFuncs.equal_range(*it);
+		static constexpr auto compareById = [](auto const* a, auto const* b) { return a->id() < b->id(); };
+		std::set<FunctionDefinition const*, decltype(compareById)> baseFunctions(compareById);
+		for (auto nextSignature = nonOverriddenFunctions.upper_bound(*it); it != nextSignature; ++it)
+			baseFunctions.insert(*it);
 
-		// Only one function
-		if (next(begin) == end)
+		if (baseFunctions.size() <= 1)
 			continue;
 
-		// Not an overridable function
-		if ((*it)->isConstructor())
+		// Construct the override graph for this signature.
+		// Reserve node 0 for the current contract and node
+		// 1 for an artificial top node to which all override paths
+		// connect at the end.
+		struct OverrideGraph
 		{
-			for (begin++; begin != end; begin++)
-				solAssert((*begin)->isConstructor(), "All functions in range expected to be constructors!");
-			continue;
-		}
+			OverrideGraph(decltype(baseFunctions) const& _baseFunctions)
+			{
+				for (auto const* baseFunction: _baseFunctions)
+					addEdge(0, visit(baseFunction));
+			}
+			std::map<FunctionDefinition const*, int> nodes;
+			std::map<int, FunctionDefinition const*> nodeInv;
+			std::map<int, std::set<int>> edges;
+			int numNodes = 2;
+			void addEdge(int _a, int _b)
```

Comment:
Any reason why the graph is undirected? Wouldn't it be more precise to have a DAG?

Reply:
Well, the cut vertex detection algorithm expects undirected graphs to work (DFS for directed graphs will be very different when compared to undirected graphs, I'm not sure the same algorithm will work...). And a directed graph has a cut vertex iff its undirected version has a cut vertex, so it doesn't really matter... if you manage to adjust the algorithm to work on directed graphs and the result is simpler, we can make it directed, otherwise I'd just stay with the undirected version.

Reply:
Ok!